Asimily Insight Integration

amelia-asimily opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
A few sentences describing the overall goals of the pull request's commits.

## Must have
- [ ] Tests
- [ ] Documentation

amelia-asimily committed:
Initial Commit: Asimily Insight

amelia-asimily committed:
Merge branch 'contrib/amelia-asimily_asimilly' into asimilly

amelia-asimily committed:
Add Prefix Asimily

amelia-asimily committed:
Second Draft after changing CLI commands to return context output

amelia-asimily committed:
Merge branch 'demisto:master' into asimilly

amelia-asimily committed:
Add unit test for script

amelia-asimily committed:
Fix validation issues

amelia-asimily committed:
Merge branch 'master' into asimilly

amelia-asimily committed:
Update README.md

Fix graph path

amelia-asimily committed:
Update README.md

Sync README

amelia-asimily committed:
Update README.md

amelia-asimily committed:
Update README.md

amelia-asimily committed:
Update README.md

amelia-asimily committed:
Update README.md

amelia-asimily committed:
Merge branch 'demisto:master' into asimilly

amelia-asimily committed:
Fix according to pre-commit checks

amelia-asimily committed:
Merge branch 'contrib/amelia-asimily_asimilly-2' into asimilly

ilappe commented:
Hi @amelia-asimily 
Sorry for the delay 
i'm reviewing your great effort!
you can run `demisto-sdk pre-commit` command in your repo, and this will run the unit test with the required modules, correctly
let me know if you need some additional help
thanks again!

amelia-asimily committed:
Merge branch 'contrib/amelia-asimily_asimilly-2' into asimilly

amelia-asimily committed:
Merge branch 'contrib/amelia-asimily_asimilly-2' into asimilly

amelia-asimily committed:
Final fix of unit test

amelia-asimily committed:
Merge branch 'contrib/amelia-asimily_asimilly-2' into asimilly

amelia-asimily committed:
Merge branch 'contrib/amelia-asimily_asimilly-2' into asimilly

amelia-asimily commented:
Hi @ilappe I have fixed the unit test failure and coverage issue.
Can you help review the updated changes. Thanks!

ilappe commented:
> Hi @ilappe I have fixed the unit test failure and coverage issue. Can you help review the updated changes. Thanks!

Hi @amelia-asimily 
Sure
thanks a lot for your effort
i'm reviewing it..

ilappe commented on Packs/Asimily_Insight/Scripts/AsimilyPreProcessDedup/AsimilyPreProcessDedup.yml:
```diff
@@ -0,0 +1,22 @@
+runas: DBotWeakRole
+engineinfo: {}
+tests:
+- No test - this is a preprocessing script
```
```suggestion
- No tests (auto formatted)
```

amelia-asimily replied:
Updated.

ilappe commented on Packs/Asimily_Insight/Integrations/AsimilyInsight/AsimilyInsight.py:
please use the `return_results` in the main 
please see https://xsoar.pan.dev/docs/integrations/code-conventions#command-functions

amelia-asimily replied:
Updated the code to only use `return_results` in `main`.

ilappe commented on Packs/Asimily_Insight/Scripts/AsimilyPreProcessDedup/AsimilyPreProcessDedup.py:
```diff
@@ -0,0 +1,33 @@
+            return
+
+        if incident_type not in ["Asimily Anomaly", "Asimily CVE"]:
+            demisto.results(True)
```
avoid using of `demisto.results`
Use `return_results` instead
see - https://xsoar.pan.dev/docs/integrations/code-conventions#deprecated---demistoresults

please do it in all the places

amelia-asimily replied:
Fixed in all places.

ilappe reviewed (changes requested):
Great work
please see my review comments on the pack/Integration/script stuff

Benimanela reviewed (changes requested):
Hi @amelia-asimily,  
Thank you for the work on this contribution. The implementation looks good overall.
Below are some points from my security review.

### General

- Please run `demisto-sdk format` on all pack files to ensure consistent formatting.

### Incident Fields

- Please avoid creating new fields when suitable common fields already exist. For example, use the common field `Device Id` instead of introducing a new `Asimily Device ID` field.

### Incident Types

- You configured indicator extraction but did not specify any indicator types. Please either define which indicator types should be extracted or set it to extract all indicators from all fields. Refer to: https://xsoar.pan.dev/docs/incidents/incident-types

### Classifier / Mapper

- Classifier and mapper files are missing. Please create these so that users can view correctly mapped data from the fetched incidents in the layout. Refer to: https://xsoar.pan.dev/docs/incidents/incident-classification-mapping

### Playbook

- What is the intended use case for this playbook? It currently contains no steps and appears incomplete. Please provide additional details or complete the playbook logic.

Let me know once the fixes are in place so I can take another look.

amelia-asimily commented on Packs/Asimily_Insight/Integrations/AsimilyInsight/AsimilyInsight.yml:
```diff
@@ -0,0 +1,500 @@
+    - description: Query only Asimily Assets with these tags.
+      isArray: true
+      name: deviceTag
+    - auto: PREDEFINED
```
Added.

amelia-asimily commented on Packs/Asimily_Insight/Integrations/AsimilyInsight/AsimilyInsight.yml:
```diff
@@ -0,0 +1,500 @@
+    description: Fetch Asimily Asset information.
+    name: asimily-get-assetdetails
+    outputs:
+    - contextPath: AsimilyInsight.Asset.asimilydevicemacaddress
```
So the field names here must match whatever field names returned in the CommandResults object right? So these values are in align with the field names in the CommandResults object elements.

ilappe replied:
Not need to update, thanks

amelia-asimily committed:
Fix according to review

amelia-asimily committed:
Change according to review

amelia-asimily committed:
Remove indicator extraction for unused fields

amelia-asimily committed:
Complete playbook

amelia-asimily committed:
Merge branch 'contrib/amelia-asimily_asimilly-2' into asimilly

amelia-asimily committed:
Fix playbook

amelia-asimily commented on Packs/Asimily_Insight/Integrations/AsimilyInsight/AsimilyInsight.yml:
```diff
@@ -0,0 +1,500 @@
+      description: Maximum amount of items to fetch.
+      name: limit
+    description: Fetch Asimily Asset information.
+    name: asimily-get-assetdetails
```
Updated.

amelia-asimily commented on Packs/Asimily_Insight/Integrations/AsimilyInsight/AsimilyInsight.yml:
```diff
@@ -0,0 +1,500 @@
+    - contextPath: AsimilyInsight.Asset.asimilydeviceipv4address
+      description: Asimily Device IP Address.
+      type: unknown
+    - contextPath: AsimilyInsight.Asset.asimilydevicemacaddress
```
Fixed.

amelia-asimily commented on Packs/Asimily_Insight/Integrations/AsimilyInsight/AsimilyInsight.py:
Added to also return in raw_response.

ilappe replied:
The `raw_response` should be the raw response, the same as it comes from the `client` from the vendor
as in case of errors/investigations - it help to understanding the issues

amelia-asimily replied:
Updated to directly return API request raw response.

amelia-asimily commented on Packs/Asimily_Insight/Integrations/AsimilyInsight/AsimilyInsight_description.md:
```diff
@@ -0,0 +1,8 @@
+- Brief information about how to retrieve the API key of your product
+- Other useful information on how to configure your integration in XSOAR
+
+Since this is a Markdown file, we encourage you to use MD formatting for sections, sub-sections, lists, etc.
```
Updated the README.

ilappe reviewed (commented):
Hi @melia-asimily
thanks for your effort
please see a few remain comments 
thanks in advance!

amelia-asimily committed:
Merge branch 'contrib/amelia-asimily_asimilly-2' into asimilly

amelia-asimily committed:
Fix typos and review comments

amelia-asimily commented on Packs/Asimily_Insight/Integrations/AsimilyInsight/AsimilyInsight.yml:
```diff
@@ -0,0 +1,500 @@
+  section: Collect
+  additionalinfo: The maximum number of incidents to fetch each time.
+  required: false
+- name: first_fetch
```
Added.

amelia-asimily commented on Packs/Asimily_Insight/Integrations/AsimilyInsight/AsimilyInsight.yml:
```diff
@@ -0,0 +1,501 @@
+  name: longRunning
+  required: false
+  section: Collect
+  type: 8
```
Removed.

amelia-asimily committed:
Merge branch 'contrib/amelia-asimily_asimilly-2' into asimilly

amelia-asimily committed:
Update playbook image

amelia-asimily committed:
Update README.md

Update pack README per review comments.

amelia-asimily committed:
Update AsimilyInsight_description.md

amelia-asimily committed:
Update README.md

amelia-asimily committed:
Update AsimilyInsight_description.md

amelia-asimily committed:
Update AsimilyInsight_description.md

amelia-asimily committed:
Update README.md

amelia-asimily commented on Packs/Asimily_Insight/README.md:
```diff
@@ -0,0 +1,276 @@
+Integrate Asimily Insight to ingest security anomalies, CVEs, and leverage detailed asset data for streamlined incident investigation.
```
Updated the README.

amelia-asimily commented:
Hi @ilappe I have pushed in the changes for all review comments unresolved.
Can you help check review again? Thank you!

amelia-asimily committed:
Update README.md

amelia-asimily committed:
Update README.md

amelia-asimily committed:
Update AsimilyInsight_description.md

ilappe commented:
> Hi @ilappe I have pushed in the changes for all review comments unresolved. Can you help check review again? Thank you! Also I am seeing the pre-commit check fails because of the warning [IF115]. There are 3 custom incident fields I added to be searchable. Having 2-3 searchable fields is fine?

yes, you can use the .`pack-ignore` file for ignoring this warning for those fields.

the code lgfm 
let's wait for the security review, by @Benimanela

Benimanela commented:
> Hi @Benimanela
> 
> 1. I have committed all changes and used the `demisto-sdk format` command to format the files.
> 2. Regarding incident fields, besides creating the Asimily custom fields, we are also mapping fields to existing common fields to make them searchable (e.g. `devicelocalip`, `macaddress`, `devicemodel`, `hostnames`). For deviceid I will add change to map it to system field deviceid. But is it recommended/required to just remove the those asimily fields if it's mapped to some custom fields as well?
> 3. I have removed the unused extraction rules for the unrelated incident fields.
> 4. The Asimily Insight Integration is designed in a way to call the Asimily Insight APIs to retrieve data and create incidents. Classification is not requried as code internally will create different type of incidents depending on the APIs getting called. And for different incident types, the API response structure is different and mapping rules are a bit different, so the mapping is done in the code level and I didn't create a mapper as well. Is this something mandatory to have a classifier and mapper?
> 5. As for the playbook, the playbook added in the integration is just a very simple demo to showcase customer that they can use playbook to call some CLI commands provided in the Integration to gain some additional information regarding related assets when investigating the incident tickets. This is one use case our customer expected. I have added a "Done" section and 2 tasks inside the playbook.
> 
> Can you take a look at my comments and instruct if any changes still needed? Thank you!

Hi @amelia-asimily,

Sorry for the late response and thank you for addressing my previous comments!

- **General**
  - Regarding your question: If a custom Asimily field is mapped 1:1 to a system/common field and doesn’t provide additional value, it’s recommended to remove the redundant custom field to keep the integration clean and maintainable.

- **Mapper and Classifier**
  - According to XSOAR standards, both a **mapper** and a **classifier** are required:
    - The **classifier** determines the incident type based on the incoming API response.
    - The **mapper** translates API fields into user-visible incident fields.
    - Without these files, incidents can’t be properly displayed in layouts or mapped for searching/filtering.
  - Please refer to the official documentation for more details:  
    https://xsoar.pan.dev/docs/incidents/incident-classification-mapping
  - Even if your code handles some logic internally, these files are still necessary for expected platform functionality.

- **Playbook**
  - One suggestion: the first step of the playbook should check if the Asimily integration is enabled to avoid errors if the integration is disabled.

Let me know if you need further clarification or examples for creating the classifier and mapper files. You can reach me on DFIR Slack. Once you’ve addressed the points above, I’ll be happy to review again!

amelia-asimily commented:
> > Hi @Benimanela
> > 
> > 1. I have committed all changes and used the `demisto-sdk format` command to format the files.
> > 2. Regarding incident fields, besides creating the Asimily custom fields, we are also mapping fields to existing common fields to make them searchable (e.g. `devicelocalip`, `macaddress`, `devicemodel`, `hostnames`). For deviceid I will add change to map it to system field deviceid. But is it recommended/required to just remove the those asimily fields if it's mapped to some custom fields as well?
> > 3. I have removed the unused extraction rules for the unrelated incident fields.
> > 4. The Asimily Insight Integration is designed in a way to call the Asimily Insight APIs to retrieve data and create incidents. Classification is not requried as code internally will create different type of incidents depending on the APIs getting called. And for different incident types, the API response structure is different and mapping rules are a bit different, so the mapping is done in the code level and I didn't create a mapper as well. Is this something mandatory to have a classifier and mapper?
> > 5. As for the playbook, the playbook added in the integration is just a very simple demo to showcase customer that they can use playbook to call some CLI commands provided in the Integration to gain some additional information regarding related assets when investigating the incident tickets. This is one use case our customer expected. I have added a "Done" section and 2 tasks inside the playbook.
> > 
> > Can you take a look at my comments and instruct if any changes still needed? Thank you!
> 
> Hi @amelia-asimily,
> 
> Sorry for the late response and thank you for addressing my previous comments!
> 
> * **General**
>   
>   * Regarding your question: If a custom Asimily field is mapped 1:1 to a system/common field and doesn’t provide additional value, it’s recommended to remove the redundant custom field to keep the integration clean and maintainable.
> * **Mapper and Classifier**
>   
>   * According to XSOAR standards, both a **mapper** and a **classifier** are required:
>     
>     * The **classifier** determines the incident type based on the incoming API response.
>     * The **mapper** translates API fields into user-visible incident fields.
>     * Without these files, incidents can’t be properly displayed in layouts or mapped for searching/filtering.
>   * Please refer to the official documentation for more details:
>     https://xsoar.pan.dev/docs/incidents/incident-classification-mapping
>   * Even if your code handles some logic internally, these files are still necessary for expected platform functionality.
> * **Playbook**
>   
>   * One suggestion: the first step of the playbook should check if the Asimily integration is enabled to avoid errors if the integration is disabled.
> 
> Let me know if you need further clarification or examples for creating the classifier and mapper files. You can reach me on DFIR Slack. Once you’ve addressed the points above, I’ll be happy to review again!

Hi @Benimanela I haven't joined the DFIR Slack channel. Can you share a bit more about how to join thanks!

qfortress-dev commented:
> > > Hi @Benimanela
> > > 
> > > 1. I have committed all changes and used the `demisto-sdk format` command to format the files.
> > > 2. Regarding incident fields, besides creating the Asimily custom fields, we are also mapping fields to existing common fields to make them searchable (e.g. `devicelocalip`, `macaddress`, `devicemodel`, `hostnames`). For deviceid I will add change to map it to system field deviceid. But is it recommended/required to just remove the those asimily fields if it's mapped to some custom fields as well?
> > > 3. I have removed the unused extraction rules for the unrelated incident fields.
> > > 4. The Asimily Insight Integration is designed in a way to call the Asimily Insight APIs to retrieve data and create incidents. Classification is not requried as code internally will create different type of incidents depending on the APIs getting called. And for different incident types, the API response structure is different and mapping rules are a bit different, so the mapping is done in the code level and I didn't create a mapper as well. Is this something mandatory to have a classifier and mapper?
> > > 5. As for the playbook, the playbook added in the integration is just a very simple demo to showcase customer that they can use playbook to call some CLI commands provided in the Integration to gain some additional information regarding related assets when investigating the incident tickets. This is one use case our customer expected. I have added a "Done" section and 2 tasks inside the playbook.
> > > 
> > > Can you take a look at my comments and instruct if any changes still needed? Thank you!
> > 
> > 
> > Hi @amelia-asimily,
> > Sorry for the late response and thank you for addressing my previous comments!
> > 
> > * **General**
> >   
> >   * Regarding your question: If a custom Asimily field is mapped 1:1 to a system/common field and doesn’t provide additional value, it’s recommended to remove the redundant custom field to keep the integration clean and maintainable.
> > * **Mapper and Classifier**
> >   
> >   * According to XSOAR standards, both a **mapper** and a **classifier** are required:
> >     
> >     * The **classifier** determines the incident type based on the incoming API response.
> >     * The **mapper** translates API fields into user-visible incident fields.
> >     * Without these files, incidents can’t be properly displayed in layouts or mapped for searching/filtering.
> >   * Please refer to the official documentation for more details:
> >     https://xsoar.pan.dev/docs/incidents/incident-classification-mapping
> >   * Even if your code handles some logic internally, these files are still necessary for expected platform functionality.
> > * **Playbook**
> >   
> >   * One suggestion: the first step of the playbook should check if the Asimily integration is enabled to avoid errors if the integration is disabled.
> > 
> > Let me know if you need further clarification or examples for creating the classifier and mapper files. You can reach me on DFIR Slack. Once you’ve addressed the points above, I’ll be happy to review again!
> 
> Hi @Benimanela I haven't joined the DFIR Slack channel. Can you share a bit more about how to join thanks!

@edik24, could you assist?

amelia-asimily committed:
Merge branch 'contrib/amelia-asimily_asimilly-2' into asimilly

amelia-asimily committed:
Merge branch 'contrib/amelia-asimily_asimilly-2' into asimilly

amelia-asimily committed:
Merge branch 'demisto:master' into asimilly

amelia-asimily committed:
Merge branch 'contrib/amelia-asimily_asimilly-2' into asimilly

amelia-asimily committed:
Merge branch 'demisto:master' into asimilly

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=40201) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=40201) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: eyalpalo<br/>:white_check_mark: amelia-asimily<br/>:x: OmriItzhak<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=40201) it.</sub>

amelia-asimily committed:
Add classifier and mapper; fix according to review; use system searchable fields; add hostname to incident name

amelia-asimily committed:
Merge branch 'demisto:master' into asimilly

amelia-asimily committed:
Update AsimilyInsight_description.md to add classifier and mapper

amelia-asimily committed:
Update playbook screenshot

amelia-asimily commented:
@Benimanela 
Thank you for reviewing my code and giving the suggestions.
1. I have only kept `asimilyanomalyname` and `asimilycvename` two custom searchable fields.
2. I have added default classifier and mapper for the integeration. But since my code is internally handling the mapping, I didn't add the `rawJSON` field for incident creation.
3. I have updated the playbook as suggested.

Can you pls review and let me know if anything further needs to be changed? Thank you!

Benimanela reviewed (approved):
Looks good!
approved from my end

amelia-asimily committed:
Merge branch 'demisto:master' into asimilly

amelia-asimily commented:
Hi @merit-maita @ilappe @Benimanela
So just want to ask what would be the next steps here for getting the Pack released?
I tried to sync the repo with master and somehow it bring in some other people as reviewers. 
Thanks!

merit-maita commented:
hello @amelia-asimily  let's set a demo please with me and @Benimanela 
feel free to contact me over slack DFIR to set a slot

merit-maita commented:
here're our emails:
mmaayta@paloaltonetworks.com
Benim@qmasters.co

amelia-asimily committed:
Merge branch 'demisto:master' into asimilly

amelia-asimily committed:
Supress warning and add new customer anomaly id field

amelia-asimily committed:
Merge branch 'demisto:master' into asimilly

amelia-asimily committed:
Fix description and add missing new field file

amelia-asimily committed:
Fix playbook warning

amelia-asimily commented:
Hi @Benimanela @merit-maita I have pushed in the fix for the playbook issue we discussed earlier.
Can you help check and see if anything else is needed for moving to the next step? Thanks!

amelia-asimily committed:
Fix playbook error
